File: 25-reverse-nodes-in-k-group/reverse-nodes-in-k-group.cs
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     public int val;
 *     public ListNode next;
 *     public ListNode(int val=0, ListNode next=null) {
 *         this.val = val;
 *         this.next = next;
 *     }
 * }
 */
public class Solution {
    public ListNode ReverseKGroup(ListNode head, int k) {
        if (head == null || k == 1) return head;

        // Dummy node to handle the new head position
        ListNode dummy = new ListNode(0);
        dummy.next = head;
        ListNode cur = dummy, nex = dummy, pre = dummy;

        // Count the number of nodes in the list
        int count = 0;
        while (cur.next != null) {
            cur = cur.next;
            count++;
        }

        // Perform the reverse in groups of k
        while (count >= k) {
            cur = pre.next;
            nex = cur.next;
            for (int i = 1; i < k; i++) {
                cur.next = nex.next;
                nex.next = pre.next;
                pre.next = nex;
                nex = cur.next;
            }
            pre = cur;
            count -= k;
        }

        return dummy.next;
    }
}

File: 25-reverse-nodes-in-k-group/reverse-nodes-in-k-group.js
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} head
 * @param {number} k
 * @return {ListNode}
 */
var reverseKGroup = function(head, k) {
    if (!head || k===1) return head;
    let dummy=new ListNode(0);
    dummy.next=head;
    let prevGroupEnd=dummy, current=head;
    while (true){
        let kthNode=getKthNode(current, k);
        if (!kthNode) break;
        let nextGroupStart=kthNode.next, groupStart=current;
        reverse(groupStart,nextGroupStart);
        prevGroupEnd.next=kthNode;
        groupStart.next=nextGroupStart;
        prevGroupEnd=groupStart;
        current=nextGroupStart;
    }return dummy.next;
    function getKthNode(node,k){
        while (node && k>1){node=node.next; k--;}return node;
    }
    function reverse(start, end){
        let prev=null, current=start;
        while (current !==end){
            let tempNext=current.next;
            current.next=prev;
            prev=current;
            current=tempNext;
        }
    }
    
};